### DISCUSSION TITLE
Using instance() break the model

### DISCUSSION THREAD

**Discussion by Makio64** [donmccurdy/glTF-Transform#1115] [Q&A] [answered]
When using instance() the model is broken.
I was expecting instance working out of the box with threejs GLTFLoader but maybe it should be handle manually ?

Using instance() on the model : 
the instanced are in position 0,0,0
the instanced scale are incorrect

**To Reproduce**
Model to test : https://www.dropbox.com/scl/fi/kazr5j60bv3fueydfou64/suitcases.glb?rlkey=b3gnqfs3yf60c1hslj9g7sqp6&dl=0
```
await document.transform(flatten())
await document.transform(dedup())
await document.transform(instance({ min: 2 }))
```
Load with threejs GLTFLoder ( for example : https://gltf-viewer.donmccurdy.com/ )

**Expected behavior**
Instanced have correct position / scale / rotate / color.

**Versions:**
 - Version: Latest

**Comment by donmccurdy** [comment, marked as answer]
I'm guessing you my not have installed any extensions in glTF Transform. Do your console logs show anything like the error below?

> Some extensions were not registered for I/O, and will not be written.

This means the `instance()` command added glTF extensions to the scene, but the I/O class doesn't recognize those extensions.

To add support for all official Khronos glTF extensions, and [a few others](https://gltf-transform.dev/extensions):

```js
import { NodeIO } from '@gltf-transform/core';
import { ALL_EXTENSIONS } from '@gltf-transform/extensions';
import { flatten, dedup, instance } from '@gltf-transform/functions';


const io = new NodeIO().registerExtensions(ALL_EXTENSIONS);
const document = await io.read('./suitcases.glb');

await document.transform(
	flatten(),
	dedup(),
	instance({min: 2}),
);

await io.write('./suitcases-opt.glb', document);
```

A few extensions, currently [KHR_draco_mesh_compression](https://gltf-transform.dev/modules/extensions/classes/KHRDracoMeshCompression) and [EXT_meshopt_compression](https://gltf-transform.dev/modules/extensions/classes/EXTMeshoptCompression), require some additional steps to use.

**Comment by Makio64** [comment]
Thanks it worked !

But maybe called instance() or other functions should automatically register the related extensions if they are missing ?

I think it will be a nice behavior to reduce errors.

Another related question : is there any side effects register ALL_EXTENSIONS ?

**Reply by donmccurdy** [reply]
Perhaps, or throwing an error. Often you'll be reading a model that could have its own extensions, and those extensions need to be registered for I/O before reading anyway. There shouldn't be any downside to registering ALL_EXTENSIONS other than adding 25kb (minzipped) to the bundle.

**Reply by Makio64** [reply]
From what in understand adding extension when using a function which require it (instance as example) seems easier than delete unused extensions as many model might have unrecognized extensions ?

Also a function to remove unused ext seems hard to implements as we might cross unrecognized one ?

For users like me who aren't experts in the format, if gltf-transform simplifies the process it'll be appreciate. The less we have to think about the technicalities, the better! :)

**Reply by donmccurdy** [reply]
Oh, I don't mean that you need to remove unknown extensions! Just that you probably **should** register `ALL_EXTENSIONS`, unless you know your models won't need them and are trying to reduce bundle size.

I can't do that automatically for the I/O classes – the separation between the /core and /extensions packages is pretty important, and allows people to implement their own custom glTF extensions. :)
